cli/src/codecs.js

```diff
@@ -0,0 +1,363 @@
+  }
+  if (target_width && target_height) {
+    return { width: target_width, height: target_height };
+  }
+  if (!target_width) {
+    return {
+      width: Math.round((input_width / input_height) * target_height),
+      height: target_height,
+    };
+  }
+  if (!target_height) {
+    return {
+      width: target_width,
+      height: Math.round((input_height / input_width) * target_width),
+    };
+  }
+}
+
+export const preprocessors = {
+  resize: {
+    name: 'Resize',
+    description: 'Resize the image before compressing',
+    instantiate: async () => {
+      await resizePromise;
+      return (
+        buffer,
+        input_width,
+        input_height,
+        { width, height, method, premultiply, linearRGB },
+      ) => {
+        ({ width, height } = resizeWithAspect({
+          input_width,
+          input_height,
+          target_width: width,
+          target_height: height,
+        }));
+        return new ImageData(
+          resize.resize(
+            buffer,
+            input_width,
+            input_height,
+            width,
+            height,
+            resizeNameToIndex(method),
+            premultiply,
+            linearRGB,
+          ),
+          width,
+          height,
+        );
+      };
+    },
+    defaultOptions: {
+      method: 'lanczos3',
+      fitMethod: 'stretch',
+      premultiply: true,
+      linearRGB: true,
+    },
+  },
+  // // TODO: Need to handle SVGs and HQX
+  quant: {
+    name: 'ImageQuant',
+    description: 'Reduce the number of colors used (aka. paletting)',
+    instantiate: async () => {
+      const imageQuant = await imageQuantPromise;
+      return (buffer, width, height, { numColors, dither }) =>
+        new ImageData(
+          imageQuant.quantize(buffer, width, height, numColors, dither),
+          width,
+          height,
+        );
+    },
+    defaultOptions: {
+      numColors: 255,
+      dither: 1.0,
+    },
+  },
+  rotate: {
+    name: 'Rotate',
+    description: 'Rotate image',
+    instantiate: async () => {
+      return async (buffer, width, height, { numRotations }) => {
+        const degrees = (numRotations * 90) % 360;
+        const sameDimensions = degrees == 0 || degrees == 180;
+        const size = width * height * 4;
+        const { instance } = await WebAssembly.instantiate(
+          await fsp.readFile(pathify(rotateWasm)),
+        );
+        const { memory } = instance.exports;
+        const additionalPagesNeeded = Math.ceil(
+          (size * 2 - memory.buffer.byteLength + 8) / (64 * 1024),
+        );
+        if (additionalPagesNeeded > 0) {
+          memory.grow(additionalPagesNeeded);
+        }
+        const view = new Uint8ClampedArray(memory.buffer);
+        view.set(buffer, 8);
+        instance.exports.rotate(width, height, degrees);
+        return new ImageData(
+          new Uint8ClampedArray(view.slice(size + 8, size * 2 + 8)),
```

Comment:
```suggestion
          view.slice(size + 8, size * 2 + 8),
```

`.slice` result is already a copy, no need to wrap again.